File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Protocol/Identity.cs
using System.Text.RegularExpressions;

namespace Depz.Sensor.Protocol;

public enum SensorType
{
    Sr04,

    /// <summary>
    /// ToF ranger. Covers both silicon variants — VL53L8CX (base) and VL53L8CH
    /// (CX + CNH). The GET_NAME_ACTIVE_SOFTWARE probe string ("APP_VL53L8…") does
    /// not distinguish them; the CX/CH split surfaces in the decode layer as
    /// <see cref="Vl53l8.Vl53l8Variant"/> and in the USB PID hint (CH = 0xED40).
    /// </summary>
    Vl53l8,

    /// <summary>
    /// VL53L4CD single-zone ToF ranger behind the thin I2C register bridge
    /// (probe string "APP_VL53L4…"; decode layer in <c>Depz.Sensor.Vl53l4</c>).
    /// </summary>
    Vl53l4,

    Bno086,
    Unknown,
}

/// <summary>
/// Classified GET_NAME_ACTIVE_SOFTWARE string (contracts/02 §4).
/// <see cref="SensorType"/> is null in bootloader mode.
/// </summary>
public sealed record Identity(string Mode, SensorType? SensorType, string SoftwareName, string Version);

public static class IdentityParser
{
    private static readonly Regex VersionRe = new(@"_v(\d+(?:\.\d+)*)$", RegexOptions.Compiled);

    private static readonly (string Token, SensorType Sensor)[] ProductTokens =
    {
        ("SR04", Protocol.SensorType.Sr04),
        ("VL53L8", Protocol.SensorType.Vl53l8),
        ("VL53L4", Protocol.SensorType.Vl53l4),
        ("BNO086", Protocol.SensorType.Bno086),
    };

    /// <summary>
    /// Classify a GET_NAME_ACTIVE_SOFTWARE string. The string must already be
    /// stripped of trailing NUL/0xFF (<see cref="Common.StripDeviceString"/>).
    /// </summary>
    public static Identity ParseSoftwareName(string name)
    {
        Match m = VersionRe.Match(name);
        string version = m.Success ? m.Groups[1].Value : "";

        if (name.StartsWith("BOOTDEPZ", StringComparison.Ordinal))
            return new Identity("bootloader", null, name, version);

        if (name.StartsWith("APP_", StringComparison.Ordinal))
        {
            foreach (var (token, sensor) in ProductTokens)
            {
                if (name.Contains(token, StringComparison.Ordinal))
                    return new Identity("app", sensor, name, version);
            }
            return new Identity("app", Protocol.SensorType.Unknown, name, version);
        }

        return new Identity("unknown", null, name, version);
    }
}

File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/protocol/Identity.java
package ai.depz.sensor.protocol;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Firmware-name parsing (contracts/02_COMMON_COMMANDS.md §4).
 *
 * @param mode       "app" | "bootloader" | "unknown"
 * @param sensorType {@code null} in bootloader/unknown mode
 * @param softwareName the classified name
 * @param version    "" when not parseable
 */
public record Identity(String mode, SensorType sensorType, String softwareName, String version) {

    public enum SensorType {
        SR04("sr04"),
        VL53L4("vl53l4"),
        VL53L8("vl53l8"),
        BNO086("bno086"),
        UNKNOWN("unknown");

        public final String label;
        SensorType(String label) { this.label = label; }
    }

    private static final Pattern VERSION_RE = Pattern.compile("_v(\\d+(?:\\.\\d+)*)$");

    private static final Object[][] PRODUCT_TOKENS = {
        {"SR04", SensorType.SR04},
        {"VL53L4", SensorType.VL53L4},
        {"VL53L8", SensorType.VL53L8},
        {"BNO086", SensorType.BNO086},
    };

    /**
     * Classify a GET_NAME_ACTIVE_SOFTWARE string. The string must already be
     * stripped of trailing NUL/0xFF ({@link Common#stripDeviceString}).
     */
    public static Identity parseSoftwareName(String name) {
        Matcher m = VERSION_RE.matcher(name);
        String version = m.find() ? m.group(1) : "";
        if (name.startsWith("BOOTDEPZ")) {
            return new Identity("bootloader", null, name, version);
        }
        if (name.startsWith("APP_")) {
            for (Object[] tok : PRODUCT_TOKENS) {
                if (name.contains((String) tok[0])) {
                    return new Identity("app", (SensorType) tok[1], name, version);
                }
            }
            return new Identity("app", SensorType.UNKNOWN, name, version);
        }
        return new Identity("unknown", null, name, version);
    }
}
